### PR TITLE
Expose cluster-state role mappings in APIs

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR exposes operator-defined, cluster-state role mappings in the [Get role mappings API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-role-mapping.html). 

Cluster-state role mappings are returned with a reserved suffix `-read-only-operator-mapping`, to disambiguate with native role mappings stored in the security index. CS role mappings are also marked with a `_read_only` metadata flag. It's possible to query a CS role mapping using its name both with and without the suffix.  

CS role mappings can be viewed via the API, but cannot be modified. To clarify this, the PUT and DELETE role mapping endpoints return header warnings if native role mappings that name-clash with CS role mappings are created, modified, or deleted. 

The PR also prevents the creation or role mappings with names ending in `-read-only-operator-mapping` to ensure that CS role mappings and native role mappings can always be fully disambiguated.

Finally, the PR changes how CS role mappings are persisted in cluster-state. CS role mappings are written (and read from disk) in the `XContent` format. This format omits the role mapping's name. This means that if CS role mappings are ever recovered from disk (e.g., during a master-node restart), their names are erased. To address this, this PR changes CS role mapping serialization to persist the name of a mapping in a reserved metadata field, and recover it from metadata during serialization. This allows us to persist the name without BWC-breaks in role mapping `XContent` format. It also allows us to ensure that role mappings are re-written to cluster state in the new, name-preserving format the first time operator file settings are processed.

Depends on: https://github.com/elastic/elasticsearch/pull/114295
Relates: ES-9628

### CODE REVIEW

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    private static final ExpressionRoleMapping clusterStateMapping1 = new ExpressionRoleMapping(
+        "role-mapping-1-read-only-operator-mapping",
+        new FieldExpression("username", List.of(new FieldExpression.FieldValue("no_user"))),
+        List.of("role_1"),
+        null,
+        Map.of("uuid", "b9a59ba9-6b92-4be2-bb8d-02bb270cb3a7", "_foo", "something", "_read_only", true),
+        true
+    );
+    private static final ExpressionRoleMapping clusterStateMapping2 = new ExpressionRoleMapping(
+        "role-mapping-2-read-only-operator-mapping",
```
Comment (n1v0lg) [inline x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java:87]: I'm going with the `-read-only-operator-mapping` because it got too depressing writing the REST tests with the alternative: they don't work with spaces unless you URL-encode the path. 

If we want to make for nicer display items in Kibana, we could always render the role mapping names different there whereas we can't do anything about spaces in the ES API.

**Comment by n1v0lg** [comment]
I don't think the BWC failures are related. Bunch of synthetic source related tests with failures like:

```
MixedClusterClientYamlTestSuiteIT > test {p0=indices.create/21_synthetic_source_stored/index param - field ordering} FAILED
    java.lang.AssertionError: Failure at [indices.create/21_synthetic_source_stored:1128]: Expected a map containing
    a: expected "2" but was <2>
```

**Comment by n1v0lg** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public static boolean hasFallbackName(ExpressionRoleMapping expressionRoleMapping) {
+        return expressionRoleMapping.getName().equals(FALLBACK_NAME);
+    }
+
+    public static void removeNameFromMetadata(Map<String, Object> metadata) {
+        assert metadata instanceof HashMap<String, Object>;
+        metadata.remove(METADATA_NAME_FIELD);
+    }
+
+    public static ExpressionRoleMapping parseWithNameFromMetadata(XContentParser parser) throws IOException {
```
Comment (jfreden) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:70]: nit: Since this behaviour is specific to how the `RoleMappingMetadata` is parsed, I think it belongs in that class.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ * To address this, this class augments CS role mapping serialization to persist the name of a mapping in a reserved metadata field,
+ * and recover it from metadata during de-serialization.
+ * Storing the name in metadata allows us to persist the name without BWC-breaks in role mapping `XContent` format (as opposed to changing
+ * XContent serialization to persist the top-level name field).
+ * It also ensures that role mappings are re-written to cluster state in the new,
+ * name-preserving format the first time operator file settings are processed.
+ * {@link #copyWithNameInMetadata(ExpressionRoleMapping)}} is used to copy the name of a role mapping into its metadata field.
+ * This is called when the role mappings are read from operator file settings (at which point the correct name is still available).
+ * {@link #parseWithNameFromMetadata(XContentParser)} is used to parse a role mapping from XContent restoring the name from metadata.
+ */
```
Comment (jfreden) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:32]: I like that everything related to this fix is in one class, but I think it becomes a little hard to follow what's going on and it would be nice to not have to do a helper class if it's not needed. A short comment for each of the methods and putting them in the class where they're used would be better. 

I also think the constants belong in `RoleMappingMetadata` since that class is already responsible for representing a role mapping as xcontent.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java [hunk 4]
```diff
@@ -36,10 +39,23 @@ public TransportDeleteRoleMappingAction(
...
-        roleMappingStore.deleteRoleMapping(request, listener.safeMap(DeleteRoleMappingResponse::new));
+        roleMappingStore.deleteRoleMapping(request, listener.safeMap(found -> {
+            if (found && clusterStateRoleMapper.hasMapping(request.getName())) {
+                // Allow to delete a mapping with the same name in the native role mapping store as the file_settings namespace, but
+                // add a warning header to signal to the caller that this could be a problem.
+                HeaderWarning.addWarning(
+                    "A read only role mapping with the same name ["
+                        + request.getName()
+                        + "] has previously been defined in a configuration file. "
+                        + "The read only role mapping will not be deleted and will remain active."
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java:55]: Reading this it almost sounds like the delete failed for the native role mapping too. We don't want to add too much text here, but is there a way to say that the native one was successfully removed?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java [hunk 5]
```diff
@@ -41,19 +50,43 @@ public TransportGetRoleMappingsAction(
...
         }, listener::onFailure));
     }
+
+    private GetRoleMappingsResponse buildResponse(
+        Collection<ExpressionRoleMapping> clusterStateMappings,
+        Collection<ExpressionRoleMapping> nativeMappings
+    ) {
+        Stream<ExpressionRoleMapping> translatedClusterStateMappings = clusterStateMappings.stream().filter(roleMapping -> {
+            if (ReservedRoleMappingXContentNameFieldHelper.hasFallbackName(roleMapping)) {
+                logger.warn(
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:80]: 👍

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportClusterStateRoleMappingTranslator.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ * This class handles the following responsibilities to ensure correct processing of cluster-state role mappings:
+ * </p>
+ *
+ * <ol>
+ *   <li>Appends a reserved suffix to cluster-state role mapping names to avoid conflicts with native role mappings.</li>
+ *   <li>Marks the metadata of cluster-state role mappings with a reserved read-only flag.</li>
+ *   <li>Removes internal metadata flags used in processing (see {@link ReservedRoleMappingXContentNameFieldHelper}).</li>
+ * </ol>
+ */
+public final class TransportClusterStateRoleMappingTranslator {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportClusterStateRoleMappingTranslator.java:37]: I struggle with the naming of this class a little and I think it's because it does a lot of different things (translate, validate, mutate and inspect). 

I think these static methods belong in the actions where they're used and the constants belong in `ExpressionRoleMapping` since this is specific to the behaviour of that class.

File: qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/FileSettingsRoleMappingUpgradeIT.java [hunk 7]
```diff
@@ -106,6 +109,10 @@ public void testRoleMappingsAppliedOnUpgrade() throws IOException {
             );
             assertThat(roleMappings, is(not(nullValue())));
             assertThat(roleMappings.size(), equalTo(1));
+            assertThat(roleMappings, is(instanceOf(Map.class)));
+            @SuppressWarnings("unchecked")
+            Map<String, Object> roleMapping = (Map<String, Object>) roleMappings;
+            assertThat(roleMapping.keySet(), contains("everyone_kibana"));
```
[the remark this answers is not in the document]
Comment (jfreden) [reply qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/FileSettingsRoleMappingUpgradeIT.java:115]: Nice!

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+                          "roles": [ "role_2" ],
+                          "rules": { "field": { "username": "no_user" } }
+                       },
+                       "role-mapping-3": {
+                          "enabled": true,
+                          "roles": [ "role_3" ],
+                          "rules": { "field": { "username": "no_user" } },
+                          "metadata": {
+                             "_read_only" : { "field": 1 },
+                             "_es_reserved_role_mapping_name": { "still_ignored": true }
```
Comment (jfreden) [inline x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java:72]: Good test!

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java [hunk 9]
```diff
@@ -308,48 +321,97 @@ public void testRoleMappingsApplied() throws Exception {
...
+                TransportClusterStateRoleMappingTranslator.addReadOnlySuffix("everyone_kibana"),
+                "_everyone_kibana",
+                TransportClusterStateRoleMappingTranslator.addReadOnlySuffix("everyone_fleet"),
+                "zzz_mapping",
+                "123_mapping"
+            )
+        );
+
+        int readOnlyCount = 0;
+        // assert that cluster-state role mappings come last
```
Comment (jfreden) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java:372]: nit: do we care about this anymore? Looks like we don't validate the order.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java [hunk 10]
```diff
@@ -81,7 +82,22 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+            return Set.of();
+        }
+        final Set<ExpressionRoleMapping> mappings = getMappings();
+        if (names == null || names.isEmpty()) {
+            return mappings;
+        }
+        return mappings.stream().filter(it -> names.contains(it.getName())).collect(Collectors.toSet());
+    }
+
+    public Set<ExpressionRoleMapping> getMappings() {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java:100]: nit: I think it would be cleaner to turn this around, so that the `getMappings()` method calls the `getMappings(@Nullable Set<String> names)` with `null` instead.

**Review by jfreden — COMMENTED** [review]
Looks great! I think everything makes sense, I have a couple of nits. Good job and solid testing as always! 💯

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java [hunk 11]
```diff
@@ -41,19 +50,43 @@ public TransportGetRoleMappingsAction(
...
+            names = Set.of();
         } else {
             names = new HashSet<>(Arrays.asList(request.getNames()));
         }
-        this.roleMappingStore.getRoleMappings(names, ActionListener.wrap(mappings -> {
-            ExpressionRoleMapping[] array = mappings.toArray(new ExpressionRoleMapping[mappings.size()]);
-            listener.onResponse(new GetRoleMappingsResponse(array));
+        roleMappingStore.getRoleMappings(names, ActionListener.wrap(nativeRoleMappings -> {
+            final Collection<ExpressionRoleMapping> clusterStateRoleMappings = clusterStateRoleMapper.getMappings(
+                // if the API was queried with a reserved suffix for any of the names, we need to remove it because role mappings are
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:66]: Currently, this will fetch from the native mapping store without messing with the name (same as today's behavior) and attempt to fetch from cluster state with/without suffix.

I think that behavior still makes sense: in the unlikely event that users have a native mapping with that suffix it should still be returned via the API. Kibana will indeed get messed up in case there a clashing mappings but I think it's more important that users have a way of getting the clashing native mapping via the API so that they can delete it. 

WDYT about adding a header warning if the native mapping store returns any mappings with a reserved suffix?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java (hunk 3 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:32]: Fair! I'll move the code into `RoleMappingMetadata`

Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:32]: I _do_ want to keep `copyWithNameInMetadata` and `parseWithNameFromMetadata` in the same class since they really only make sense together. Let me know if you have strong objections.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            metadata,
+            roleMapping.isEnabled()
+        );
+    }
+
+    public static boolean hasFallbackName(ExpressionRoleMapping expressionRoleMapping) {
+        return expressionRoleMapping.getName().equals(FALLBACK_NAME);
+    }
+
+    public static void removeNameFromMetadata(Map<String, Object> metadata) {
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:65]: On second thought, I've just inlined the call to where it's used -- it's a single place where we are guaranteed a `HashMap` so might as well skip the public util method.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            roleMapping.isEnabled()
+        );
+    }
+
+    private static String getNameFromMetadata(ExpressionRoleMapping roleMapping) {
+        Map<String, Object> metadata = roleMapping.getMetadata();
+        if (metadata.containsKey(METADATA_NAME_FIELD) && metadata.get(METADATA_NAME_FIELD) instanceof String name) {
+            return name;
+        } else {
+            // This is valid the first time we recover from cluster-state: the old format metadata won't have a name stored in metadata yet
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ReservedRoleMappingXContentNameFieldHelper.java:87]: Discussed on Slack: `name` is implicitly expected by `ExpressionRoleMapping` to not be `null` both for the API layer so it's more straight-forward and robust to maintain that with a fallback name.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java (hunk 4 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportDeleteRoleMappingAction.java:55]: Good point. I've tweaked the text to hopefully be clearer now.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportClusterStateRoleMappingTranslator.java (hunk 6 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportClusterStateRoleMappingTranslator.java:37]: Agreed, I've dismantled this and moved its fields and methods, foremost to `ExpressionRoleMapping`

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java (hunk 10 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/ClusterStateRoleMapper.java:100]: 👍

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:69]: Sorry, I think I misunderstood the order of things here. 

So, if we have a `mapping-1` the native store and cluster state, and we try to get `mapping-1`, we would get [`mapping-1`, `mapping-1-read-only-operator-mapping`]. 

I think just getting `mapping-1` from the native store would make a little more sense in this case and only return `mapping-1-read-only-operator-mapping` if the full suffix is provided. I think that Kibana probably expects only a single role mapping back when they provide a single name? 

The argument could be made that a user might query for what they have in the operator defined role mappings to validate something and hence use the name from there, so I can see both sides, but if it breaks Kibana, it's probably better to only return from the native store if present in both.

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java (hunk 1 above)
Comment (n1v0lg) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java:87]: Thanks for raising this!

Fully agreed that it looks better in the Kibana role mapping page. Dev Tools however becomes very difficult to use because the path is also separated by commas. Furthermore, switching to ` (read-only)` still results in fairly confusing test failures even with URL encoding. For instance the equality checks for the PUT API don't work anymore. Handling URL encoding feels error-prone and not worth the effort.

> If you decide to not do that, I think including the original name in the metadata returned would be nice so Kibana doesn't have to do some string magic to strip the suffix.

I think the right answer might be to add a new field called `display_name` to the top-level expression mapping XContent class. I'm really hesitant to use a metadata field beyond the backend hack (which is still something we can choose to clean up) for something that really feels like a missing top-level field. Or better yet: at the XContent layer, we could add an `id` field, and let `name` become "display name". I don't think any of this (not even surfacing the metadata field in Kibana) is viable for 8.16.0 though so I propose to go with the current implementation, and then follow up, having coordinated with Kibana.

Does this sound like an okay path forward?

Comment (jfreden) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/rolemapping/RoleMappingRestIT.java:87]: Yes, that sounds good!  Thanks for exploring this path. I don't think it's worth spending too much time getting this to work. 

I like the idea of a display name eventually, but like you said, not something we should add until there is a need for it. Adding it to the metadata might force us to have it there forever, so I agree that it makes sense to not do that at this point.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:69]: So Kibana can handle returning both and will pick the first one. We discussed on slack and decided to stick with the current implementation.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java:69]: Discussed this on Slack:

1. Kibana will pick the first mapping returned for the single mapping view. Since we ensure that native mappings always come first, that's the one that will be chosen.
2. There's (some) value in allowing users to query CS role mappings with or without the suffix. 

As such, keeping this as is.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java (hunk 9 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/RoleMappingFileSettingsIT.java]: It actually matters (see my comment here: https://github.com/elastic/elasticsearch/pull/114951#discussion_r1810573906) so I've touched this up to assert order.

**Review by jfreden — APPROVED** [review]
Thanks for addressing all the comments and for the very thorough testing. Great job on this! LGTM!

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@jfreden thanks for the thorough review!
